pkg/utils/resources/resources.go

```diff
@@ -93,20 +93,94 @@ func Subtract(lhs, rhs v1.ResourceList) v1.ResourceList {
-	var resources v1.ResourceRequirements
+func PodRequests(pod *v1.Pod) v1.ResourceList {
+	reqs := v1.ResourceList{}
+
 	for _, container := range pod.Spec.Containers {
-		resources.Requests = MergeInto(resources.Requests, MergeResourceLimitsIntoRequests(container))
-		resources.Limits = MergeInto(resources.Limits, container.Resources.Limits)
+		containerReqs := container.Resources.Requests
+
+		MergeInto(reqs, containerReqs)
+
 	}
+
+	restartableInitContainerReqs := v1.ResourceList{}
+	initContainerReqs := v1.ResourceList{}
+
 	for _, container := range pod.Spec.InitContainers {
-		resources.Requests = MaxResources(resources.Requests, MergeResourceLimitsIntoRequests(container))
-		resources.Limits = MaxResources(resources.Limits, container.Resources.Limits)
+		containerReqs := container.Resources.Requests
+
+		if container.RestartPolicy != nil && *container.RestartPolicy == v1.ContainerRestartPolicyAlways {
+			MergeInto(reqs, containerReqs)
+
+			MergeInto(restartableInitContainerReqs, containerReqs)
+			containerReqs = restartableInitContainerReqs
+		} else {
+			tmp := v1.ResourceList{}
+			MergeInto(tmp, containerReqs)
+			MergeInto(tmp, restartableInitContainerReqs)
+			containerReqs = tmp
+		}
+
+		initContainerReqs = MaxResources(initContainerReqs, containerReqs)
 	}
+
+	reqs = MaxResources(reqs, initContainerReqs)
+
 	if pod.Spec.Overhead != nil {
-		resources.Requests = MergeInto(resources.Requests, pod.Spec.Overhead)
+		MergeInto(reqs, pod.Spec.Overhead)
+	}
+
+	return reqs
+}
+
+func PodLimits(pod *v1.Pod) v1.ResourceList {
+	limits := v1.ResourceList{}
+
+	for _, container := range pod.Spec.Containers {
+
+		MergeInto(limits, container.Resources.Limits)
 	}
+
+	restartableInitContainerLimits := v1.ResourceList{}
+	initContainerLimits := v1.ResourceList{}
+
+	for _, container := range pod.Spec.InitContainers {
+		containerLimits := container.Resources.Limits
+		if container.RestartPolicy != nil && *container.RestartPolicy == v1.ContainerRestartPolicyAlways {
+			MergeInto(limits, containerLimits)
+
+			MergeInto(restartableInitContainerLimits, containerLimits)
+			containerLimits = restartableInitContainerLimits
+		} else {
+			tmp := v1.ResourceList{}
+			MergeInto(tmp, containerLimits)
+			MergeInto(tmp, restartableInitContainerLimits)
+			containerLimits = tmp
+		}
+
+		initContainerLimits = MaxResources(initContainerLimits, containerLimits)
+	}
+
+	limits = MaxResources(limits, initContainerLimits)
+
+	if pod.Spec.Overhead != nil {
+		for name, quantity := range pod.Spec.Overhead {
+			if value, ok := limits[name]; ok && !value.IsZero() {
+				value.Add(quantity)
+				limits[name] = value
+			}
+		}
+	}
+
+	return limits
+}
+
+// Ceiling calculates the max between the sum of container resources and max of initContainers along with sidecar feature consideration
+func Ceiling(pod *v1.Pod) v1.ResourceRequirements {
+	var resources v1.ResourceRequirements
+
+	resources.Requests = PodRequests(pod)
```

Comment:
This change breaks our existing logic. Before we were representing pod requests using pod limits if the pod container resource requests didn't exist. I'm not necessarily convinced that we should be doing this (I think the behavior might actually be _more_ confusing for users) but I'd be interested to hear @tzneal or @ellistarn 's take on this since one or both of them wrote the original code.

Reply:
Yeah I guess ```MergeResourceLimitsIntoRequests``` makes the difference happen